### ISSUE TITLE
Supporting explicit page titles in front-matter YAML

### ISSUE THREAD

**Issue by choldgraf** [executablebooks/meta#140]
Sphinx has kind-of a wonky way of handling document titles, which follows the following rules (I think):

1. If the top-level section of a doctree begins with a `title` node, then that becomes the `{{ title }}` in the HTML context.
2. If it does not begin with a `title` node, then:
  - If the document has `title` in its docinfo, that is used as the `{{ title }}`
  - If the document has no `title` in its docinfo, then the first `title` node of the doctree becomes `{{ title }}`

Generally speaking, this doesn't really matter because Sphinx themes just render the doctree directly and most people put the title of the page *in* the doctree. E.g. they structure their documentation like this:

```md
# My title

## Header 1

My content
```

And themes don't do anything with the `{{ title }}` template variable.

However, a very common pattern in markdown (esp in SSGs) is to explicitly note the title in the page metadata. For example:

```md
---
title: My title
---

# Header 1

My content
```

I wonder if we should support such a pattern. This would have a couple of benefits:

- Users wouldn't need to reserve their top-level header level for just the document title (`#`)
- The document title could be more explicit
- We could render the document title in "special" ways (e.g. if we wanted to support an author by-line under the title, we'd want to render the `{{ title }}` variable directly in our template).

However it would need to be an optional feature, and makes me feel like we'd be departing from how Sphinx thinks about the document model.

I'm curious if folks have an opinion about this...

**Comment by akhmerov** [comment]
I think improving how titles are determined is indeed important because of how easy it is to accidentally mess up (e.g. executablebooks/jupyter-book#923).

I think specifying the title via frontmatter or even toc is a more robust pattern compared to manually checking the header levels in the document.

**Comment by choldgraf** [comment]
I don't think we can change the _default_ behavior here, but I do think we can add titles as a failsafe fallback. Otherwise we'd be breaking with how Sphinx treats titles (by default, anyway)
